File: Python/ACsN_code_Py/ACSN.py
import pickle
import os
import sys
import time
import numpy as np
from colorama import Fore, Back, Style

from ACSN_processing import ACSN_processing
from ACSN_initialization import ACSN_initialization
from ACSN_processing_parallel import ACSN_processing_parallel
from ACSN_processing_video import ACSN_processing_video
from numba import jit
import cupy as cp

# # varargin is a list with different datatypes
# # for python it will be a dictionary
# # varargin = {"Gain": xxx, "Offset": xxx, "Hotspot": xxx, "Level": xxx, 
#     # "Mode": xxx, "SaveFileName": xxx, 
#     # "Video": xxx, "Window": xxx, "Alpha": xxx, "QualityMap": xxx}

def ACSN(I,NA,Lambda,PixelSize,varargin):

    start = time.perf_counter()
    # Assumes I is a 3D variable
    Qscore = np.zeros((I.shape[2], 1))
    sigma = np.zeros((I.shape[2], 1))
    img = np.zeros(I.shape)
    Qmap = np.zeros(I.shape[0])
    I, Gain, Offset, Hotspot, Level, Mode, SaveFileName, Video, Window, alpha, QM, weight = ACSN_initialization(I, varargin=varargin)

    ## main theme

    if (Mode == "Fast"):
        img, Qmap, Qscore = ACSN_processing_parallel(I, NA, Lambda, PixelSize, Gain, Offset, Hotspot, QM, Qmap, Qscore, sigma, img, Video, weight)
    elif (Video == "yes"):
        img, Qmap, Qscore = ACSN_processing_video(I, NA, Lambda, PixelSize, Gain, Offset, Hotspot, QM, Qmap, Qscore, sigma, img, Video, weight)
    else:
        img, Qmap, Qscore = ACSN_processing(I, NA, Lambda, PixelSize, Gain, Offset, Hotspot, QM, Qmap, Qscore, sigma, img, weight)

    ## finale

    end = time.perf_counter()
    print("Elapsed Time: " + str(end - start) + " seconds" +"\n")
    print("Average Quality: ")
    Av_qI = np.mean(Qscore.flatten())
    if (Av_qI >= 0.6):
        print("High: " + str(Av_qI) + "\n")
    elif (abs(Av_qI - 0.5) < 0.1):
        print("Medium: " + str(Av_qI) + "\n")
    else:
        print("Low: " + str(Av_qI) + "\n")
    
    return Qscore, sigma, img, SaveFileName

File: Matlab/ACsN_code/ACSN.m
% ACsN v.1.5  Automatic Correction of sCMOS-related Noise
%
% SYNOPSIS:
%   [img, Qscore,elapsedTime, sigma] = ACSN(I,NA,Lambda,PixelSize,PropertyName,PropertyValue)
%
% INPUTS:
%   I
%       Noisy image: variable name or file name (only .tif files)
%   NA
%       Numerical Aperture
%   Lambda
%       Wavelength of the emitter
%   PixelSize
%       Pixel size in the image plane in micron (Camera pixel size divided by the magnification)
%
%   Properties:
%
%       Gain
%           gain map of the Camera
%       Offset
%           map of the Camera offset
%       Video
%           'yes' | 'no' | 'auto' (default)
%       Hotspot    (hotspot removal)
%           0 | 1 (default)
%       Mode
%           'Normal' | 'Fast' (default)
%       Weight
%           weight for noise estimation (decrease to reduce smoothing)
%
%
% OUTPUTS:
%   img
%       Denoised image
%   Qscore
%       Image quality score
%   sigma
%       estimated noise variation
%   elapsedTime
%       elapsed time for the denoising
%
%
% (C) Copyright 2019                Biagio Mandracchia
%     All rights reserved
%
% Biagio Mandracchia, February 2019

function [img, varargout] = ACSN(I,NA,Lambda,PixelSize,varargin) %#ok<INUSD>


%% ouverture
timerVal = tic;
img = zeros(size(I));
Qscore = zeros(size(I,3),1);
sigma = [];
Qmap = zeros(size(I)); %#ok<NASGU>
ACSN_initialization;

fprintf('ACsN v1.5\n')

%% main theme

if strcmp(Mode,'Fast')
    ACSN_processing_parallel;
elseif size(I,3)>1
    ACSN_processing_video;
else
    ACSN_processing;
end


%% finale

elapsedTime = toc(timerVal);
fprintf('\nElapsed time (s):')
disp(elapsedTime);
fprintf('Average Quality: ')
Av_qI  = mean(Qscore(:));
if Av_qI >= 0.6
    cprintf([0,0.5,0],[num2str(Av_qI) '\n\n']);
elseif abs(Av_qI - 0.5) < 0.1
    cprintf([0.75,0.5,0],[num2str(Av_qI) '\n\n']);
else
    cprintf([0.75,0,0],[num2str(Av_qI) '\n\n']);
end

out = {Qscore,elapsedTime,sigma};

for idx = 1:(nargout-1)
    varargout{idx} = out{idx}; %#ok<AGROW>
end

end
